tests/torchtune/datasets/test_packed_dataset.py

```diff
@@ -0,0 +1,86 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import pytest
+from tests.test_utils import DummyTokenizer
+from torch.utils.data import Dataset
+
+from torchtune.datasets import PackedDataset
+
+
+class DummyDataset(Dataset):
+    def __init__(self, sample_size):
+        self.sample_size = sample_size
+
+    def __getitem__(self, index):
+        if index >= 1000:
+            raise IndexError()
+        return [index] * self.sample_size, [index] * self.sample_size
+
+
+class DummyRealDataset(Dataset):
+    def __init__(self):
+        self.samples_list = [
+            "This is a packing test",
+            "A fantastic test. It should pack two samples.",
+            "This one will not be fully packed.",
+        ]
+        self.tokenizer = DummyTokenizer()
+
+    def __getitem__(self, index):
+        tokens = self.tokenizer.encode(self.samples_list[index])
+        return tokens, tokens
+
+
+class TestPackedDataset:
+    @pytest.mark.parametrize("max_seq_len", [10, 25])
+    @pytest.mark.parametrize("sample_size", [2, 5])
+    @pytest.mark.parametrize("max_rows", [5, 10])
+    @pytest.mark.parametrize("split_samples", [True, False])
+    def test_packed_dataset(self, max_seq_len, sample_size, max_rows, split_samples):
+        dataset = DummyDataset(sample_size)
+        packed = PackedDataset(
+            dataset,
+            max_seq_len=max_seq_len,
+            max_rows=max_rows,
+            split_samples=split_samples,
+        )
+        # Check we get right number of packs
+        assert len(packed) == max_rows
+        # Check input ids and labels are same length
+        assert len(packed[0][0]) == len(packed[0][1])
+        # Check that samples are packed correctly - very last individual sample
+        # should have index value of the number of times dataset was iterated over
+        if split_samples:
+            # If we split samples, we'll know how many samples by taking the
+            # full length and dividing by sample size
+            last_index, remainder = divmod(max_rows * max_seq_len, sample_size)
```

Comment:
Did not even know divmod was a thing. Nice

Reply:
Actually could also use math.ceil(max_rows * max_seq_len / sample_size) here I think? Maybe that's clearer tbh